### PR TITLE
FetchIndicatorsFromFile - Adding domain auto-detect support and default_type bug fix

### PR DESCRIPTION
**Pull request by roysagi**
## Status
- [x] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/23721

## Description
Fixed a bug where default type would not attach in case no type was detected - regardless if auto_detection was selected.

Also added support for Domain auto detection.

moved the limit and offset functionality to get better run times on long inputs

@glicht -FYI

### CODE REVIEW

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py [hunk 1]
```diff
@@ -16,19 +23,22 @@ def csv_file_to_indicator_list(file_path, col_num, starting_row, auto_detect, de
...
                 if type_col:
                     indicator_type = row[int(type_col) - 1]
 
-                elif not auto_detect:
-                    indicator_type = default_type
+                if indicator_type is None:
+                    if default_type is None:
+                        continue
+                    else:
+                        indicator_type = default_type
```
Comment (ronykoz) [inline Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py:41]: extract to another function `get_indicator_type`

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py [hunk 2]
```diff
@@ -16,19 +23,22 @@ def csv_file_to_indicator_list(file_path, col_num, starting_row, auto_detect, de
     with open(file_path) as csv_file:
         file_reader = csv.reader(csv_file)
         for row in file_reader:
-            if line_index >= starting_row:
+            if line_index >= starting_row + offset and len(row) != 0:
```
Comment (teizenman) [inline Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py:26]: is it possible to slice the `file_reader` object to start from `starting_row + offset`?

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py [hunk 3]
```diff
@@ -16,19 +23,22 @@ def csv_file_to_indicator_list(file_path, col_num, starting_row, auto_detect, de
...
 
                 indicator_type = detect_type(indicator)
 
-                if indicator_type is None:
-                    continue
+                if not auto_detect:
+                    indicator_type = default_type
 
                 if type_col:
                     indicator_type = row[int(type_col) - 1]
```
Comment (teizenman) [inline Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py:35]: are we sure that this access will not be out of range?

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py (hunk 1 above)
Comment (teizenman) [reply Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py:41]: and maybe add lines 24-25 to it, cause it seems related

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py [hunk 4]
```diff
@@ -37,10 +47,14 @@ def csv_file_to_indicator_list(file_path, col_num, starting_row, auto_detect, de
 
             line_index = line_index + 1
 
+            if limit and len(indicator_list) == int(str(limit)):
```
Comment (teizenman) [inline Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py:50]: do we need the double casting to `str` and then to `int`?

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py [hunk 5]
```diff
@@ -53,35 +67,43 @@ def xls_file_to_indicator_list(file_path, sheet_name, col_num, starting_row, aut
         xl_sheet = xl_woorkbook.sheet_by_index(0)
 
     for row_index in range(0, xl_sheet.nrows):
```
Comment (teizenman) [inline Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py:69]:
```suggestion
    for row_index in range(starting_row + offset, xl_sheet.nrows):
```

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py [hunk 6]
```diff
@@ -53,35 +67,43 @@ def xls_file_to_indicator_list(file_path, sheet_name, col_num, starting_row, aut
...
-        if row_index >= starting_row:
+        if row_index >= starting_row + offset:
```
Comment (teizenman) [inline Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py:70]:
```suggestion
```

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py [hunk 7]
```diff
@@ -53,35 +67,43 @@ def xls_file_to_indicator_list(file_path, sheet_name, col_num, starting_row, aut
...
             if type_col:
                 indicator_type = xl_sheet.cell(row_index, int(type_col) - 1).value
 
+            # indicator not recognized
+            if indicator_type is None:
+                if default_type is None:
+                    continue
+                else:
+                    indicator_type = default_type
+
```
Comment (teizenman) [inline Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py:87]: use the function from Rony's comment

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py [hunk 8]
```diff
@@ -53,35 +67,43 @@ def xls_file_to_indicator_list(file_path, sheet_name, col_num, starting_row, aut
...
             indicator_list.append({
                 'type': indicator_type,
                 'value': indicator
             })
 
+        if limit and len(indicator_list) == int(str(limit)):
```
Comment (teizenman) [inline Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py:93]: same question as above

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py [hunk 9]
```diff
@@ -106,6 +132,9 @@ def txt_file_to_indicator_list(file_path, auto_detect, default_type):
...
```
Comment (teizenman) [inline Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py:135]: last time

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py [hunk 10]
```diff
@@ -168,37 +202,18 @@ def fetch_indicators_from_file(args):
...
-
-    indicator_list_len = len(indicator_list)
+        indicator_list = txt_file_to_indicator_list(file_path, auto_detect, default_type, limit, offset)
 
-    if limit:
-        limit = int(str(limit))
-        indicator_list = indicator_list[offset: limit + offset]
-
-    human_readable = tableToMarkdown("Indicators from {}:".format(file_path), indicator_list,
+    human_readable = tableToMarkdown("Indicators from {}:".format(file_name), indicator_list,
```
Comment (teizenman) [inline Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py:214]: fstring?

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py (hunk 4 above)
Comment (roysagi) [reply Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py:40]: returns a mypy error otherwise

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py (hunk 5 above)
Comment (roysagi) [reply Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py]: done

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py (hunk 6 above)
Comment (roysagi) [reply Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py]: done

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py (hunk 8 above)
Comment (roysagi) [reply Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py:70]: same answer

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py (hunk 9 above)
Comment (roysagi) [reply Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py:109]: and again

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py (hunk 10 above)
Comment (roysagi) [reply Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py]: done

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py (hunk 1 above)
Comment (roysagi) [reply Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py]: done

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py (hunk 2 above)
Comment (roysagi) [reply Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py:27]: not that I could find in docs nor examples- https://docs.python.org/3/library/csv.html

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py (hunk 3 above)
Comment (roysagi) [reply Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py]: yes I explicitly stated in the argument description that the type_col should start from 1:
image

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py (hunk 7 above)
Comment (roysagi) [reply Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py]: done

**Comment by roysagi** [comment]
@kirbles19 - Can you please review the following:
1. [Documentation](https://github.com/demisto/content/blob/fetch_indicators_from_file_fix/Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/README.md)
2. [Changelog](https://github.com/demisto/content/blob/fetch_indicators_from_file_fix/Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/CHANGELOG.md)

File: Tests/test_content.py [hunk 11]
```diff
@@ -602,8 +603,7 @@ def run_test_scenario(tests_settings, t, proxy, default_test_timeout, skipped_te
     # Skip version mismatch test
     test_from_version = t.get('fromversion', '0.0.0')
     test_to_version = t.get('toversion', '99.99.99')
-    if (server_version_compare(test_from_version, server_numeric_version) > 0
-            or server_version_compare(test_to_version, server_numeric_version) < 0):
+    if not (LooseVersion(test_from_version) <= LooseVersion(server_numeric_version) < LooseVersion(test_to_version)):
```
Comment (anara123) [inline Tests/test_content.py:606]:
```suggestion
    if not (LooseVersion(test_from_version) <= LooseVersion(server_numeric_version) <= LooseVersion(test_to_version)):
```

**Comment by kirbles19** [comment]
@roysagi Done.

File: Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py [hunk 12]
```diff
@@ -5,9 +5,17 @@
 import re
 import xlrd
 import csv
+import tldextract
+import warnings
+import traceback
 
+# Disable insecure warnings
+requests.packages.urllib3.disable_warnings()
```
Comment (glicht) [inline Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py:13]: Why is this here?

Comment (roysagi) [reply Packs/FetchIndicatorsFromFile/Scripts/FetchIndicatorsFromFile/FetchIndicatorsFromFile.py:13]: The use of tldextract package raises SSL errors. this is used to disable these errors (same as in JSONFeedApiModule where I took the Domain recognition from)
